v2/apiserver/internal/core/events.go

```diff
@@ -537,35 +537,41 @@ func (e *eventsService) CancelMany(
 		)
 	}
 
-	events, err := e.eventsStore.CancelMany(ctx, selector)
+	eventCh, err := e.eventsStore.CancelMany(ctx, selector)
 	if err != nil {
 		return result, errors.Wrap(err, "error canceling events in store")
 	}
 
-	result.Count = int64(len(events.Items))
-
-	// TODO: This could take a while, so we don't do it synchronously. But what if
-	// the process dies while this is in-progress? Can we find a quicker, more
-	// efficient way to do this?
-	go func() {
-		for _, event := range events.Items {
-			if err := e.substrate.DeleteWorkerAndJobs(
-				context.Background(), // Deliberately not using request context
-				project,
-				event,
-			); err != nil {
-				log.Println(
-					errors.Wrapf(
-						err,
-						"error deleting event %q worker and jobs from the substrate",
-						event.ID,
-					),
-				)
+	// When an event comes in via the corresponding channel,
+	// kick off a go routine to execute cleanup.
+	for {
+		select {
+		case event, ok := <-eventCh:
+			if ok {
+				result.Count++
+				// TODO: This could take a while, so we don't do it synchronously. But
+				// what if the process dies while this is in-progress? Can we find a
+				// quicker, more efficient way to do this?
+				go func() {
```

Comment:
It looks like we have the potential to launch an unbounded number of goroutines. It's possible and practical to launch very large numbers of goroutines (I think the Go docs say tens of thousands is practical), but there's the potential here to be dealing with much larger numbers. And each of these goroutines closes over an event that was read off the channel. So if we fire up 1m goroutines, we have 1m events-- we're back to a condition that could blow out the heap.

What if we did something like this instead:

```
eventCh, err := e.eventsStore.CancelMany(ctx, selector)
if err != nil {
  return result, errors.Wrap(err, "error canceling events in store")
}
// Fan out to a finite number of goroutines
for (i := 0; i < 10; i++) {
  go func() {
    for {
      select {
        // ... cleanup ...
      }
    }
  }()
}
```

Reply:
The goroutine batch approach seems fine... though, say we have just one canceled event to cleanup after... we'll still launch 10 routines (in this example), right?  Would the other 9 routines (quickly returning, admittedly) incur any concerning overhead?

Reply:
Playing around with things now... realizing we also have a need to synchronously tally the canceled/deleted event count and hand this back to the user, which wouldn't work if we increment in a go routine.  Thinking how we might still 'batch' go routines with this in mind...

Reply:
There's an atomic counter doohickey in the atd lib's sync package, I think.

Reply:
That would segue nicely into adding a `-w/--wait` option for these commands, since we'd have to wait for all routines to finish to return the final tally.  Lemme just see if it's feasible to do so w/o waiting...

Reply:
As for whether it's overkill to launch n goroutimes to clean up just one event... we have no way of knowing headed into the loop that we're only cleaning up after m < n events. Thankfully the amount of overhead should be pretty low. The excess goroutines should idle almost immediately and then exit very shortly after that.

Also, we _could_ modify the store code to return the count of affected events along with the channel. That means going into the loop we would know how many events we expect to clean up after and we could use that info to set a floor on the number of goroutines.... it would also stop us from having to mess with an atomic counter.

Reply:
Ohhh great point re: getting the count from the store.  I like that.  That's probably the best way to go.

Reply:
Ok, updated in https://github.com/brigadecore/brigade/pull/1272/commits/dd20cdc7ff22a4bbaa3522f71fc585774bf7062f (apologies, forgot to group commits by scope).  PTAL when convenient.